src/tool_cache.rs

```diff
@@ -15,42 +15,46 @@ use crate::{
     ci_string::CiString,
     config::ToolSpec,
     fs::{self, File},
-    paths,
+    paths::ForemanPaths,
     tool_provider::ToolProvider,
 };
 
-fn index_file() -> PathBuf {
-    let mut path = paths::base_dir();
-    path.push("tool-cache.json");
-    path
-}
-
 /// Contains the current state of all of the tools that Foreman manages.
 #[derive(Debug, Default, Serialize, Deserialize)]
 pub struct ToolCache {
     pub tools: HashMap<CiString, ToolEntry>,
     #[serde(skip)]
-    providers: ToolProvider,
+    paths: ForemanPaths,
 }
 
 impl ToolCache {
     #[must_use]
-    pub fn run(tool: &ToolSpec, version: &Version, args: Vec<String>) -> i32 {
+    pub fn run(&self, tool: &ToolSpec, version: &Version, args: Vec<String>) -> i32 {
         log::debug!("Running tool {}", tool);
 
-        let mut tool_path = paths::tools_dir();
-        let exe_name = tool_identifier_to_exe_name(tool, version);
-        tool_path.push(exe_name);
+        let tool_path = self.get_tool_exe_path(tool, version);
 
-        let status = process::Command::new(tool_path)
+        let status = process::Command::new(&tool_path)
             .args(args)
             .status()
+            .map_err(|e| {
+                format!(
+                    "an error happened trying to run `{}` at `{}`: {}",
```

Comment:
When does this error occur? If there's any chance it could happen for end users, we should make sure the error is very clear about it being foreman's problem, so they don't chase ghosts in whatever tool they were trying to use

Reply:
Yes you're right, I'll precise that in the message!